services/mlmodel/mlmodel.go

```diff
@@ -26,10 +30,127 @@ func init() {
 type Service interface {
 	resource.Resource
-	Infer(ctx context.Context, input map[string]interface{}) (map[string]interface{}, error)
+	Infer(ctx context.Context, tensors ml.Tensors, input map[string]interface{}) (ml.Tensors, map[string]interface{}, error)
 	Metadata(ctx context.Context) (MLMetadata, error)
 }
 
+// TensorsToProto turns the ml.Tensors map into a protobuf message of FlatTensors.
+func TensorsToProto(ts ml.Tensors) (*servicepb.FlatTensors, error) {
+	pbts := &servicepb.FlatTensors{
+		Tensors: make(map[string]*servicepb.FlatTensor),
+	}
+	for name, t := range ts {
+		tp, err := tensorToProto(t)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to convert tensor to proto message")
+		}
+		pbts.Tensors[name] = tp
+	}
+	return pbts, nil
+}
+
+func tensorToProto(t *tensor.Dense) (*servicepb.FlatTensor, error) {
+	ftpb := &servicepb.FlatTensor{}
+	shape := t.Shape()
+	for _, s := range shape {
+		ftpb.Shape = append(ftpb.Shape, uint64(s))
+	}
+	// switch on data type of the underlying array
+	data := t.Data()
+	switch dataSlice := data.(type) {
+	case []int8:
+		unsafeByteSlice := *(*[]byte)(unsafe.Pointer(&dataSlice)) //nolint:gosec
+		data := &servicepb.FlatTensorDataInt8{}
+		data.Data = append(data.Data, unsafeByteSlice...)
+		ftpb.Tensor = &servicepb.FlatTensor_Int8Tensor{Int8Tensor: data}
+	case []uint8:
+		ftpb.Tensor = &servicepb.FlatTensor_Uint8Tensor{
+			Uint8Tensor: &servicepb.FlatTensorDataUInt8{
+				Data: dataSlice,
+			},
+		}
+	case []int16:
+		ftpb.Tensor = &servicepb.FlatTensor_Int16Tensor{
+			Int16Tensor: &servicepb.FlatTensorDataInt16{
+				Data: int16ToUint32(dataSlice),
+			},
+		}
+	case []uint16:
+		ftpb.Tensor = &servicepb.FlatTensor_Uint16Tensor{
+			Uint16Tensor: &servicepb.FlatTensorDataUInt16{
+				Data: uint16ToUint32(dataSlice),
+			},
+		}
+	case []int32:
+		ftpb.Tensor = &servicepb.FlatTensor_Int32Tensor{
+			Int32Tensor: &servicepb.FlatTensorDataInt32{
+				Data: dataSlice,
+			},
+		}
+	case []uint32:
+		ftpb.Tensor = &servicepb.FlatTensor_Uint32Tensor{
+			Uint32Tensor: &servicepb.FlatTensorDataUInt32{
+				Data: dataSlice,
+			},
+		}
+	case []int64:
+		ftpb.Tensor = &servicepb.FlatTensor_Int64Tensor{
+			Int64Tensor: &servicepb.FlatTensorDataInt64{
+				Data: dataSlice,
+			},
+		}
+	case []uint64:
+		ftpb.Tensor = &servicepb.FlatTensor_Uint64Tensor{
+			Uint64Tensor: &servicepb.FlatTensorDataUInt64{
+				Data: dataSlice,
+			},
+		}
+	case []int:
+		unsafeInt64Slice := *(*[]int64)(unsafe.Pointer(&dataSlice)) //nolint:gosec
+		data := &servicepb.FlatTensorDataInt64{}
+		data.Data = append(data.Data, unsafeInt64Slice...)
+		ftpb.Tensor = &servicepb.FlatTensor_Int64Tensor{Int64Tensor: data}
+	case []uint:
+		unsafeUint64Slice := *(*[]uint64)(unsafe.Pointer(&dataSlice)) //nolint:gosec
+		data := &servicepb.FlatTensorDataUInt64{}
+		data.Data = append(data.Data, unsafeUint64Slice...)
+		ftpb.Tensor = &servicepb.FlatTensor_Uint64Tensor{Uint64Tensor: data}
```

Comment:
Making an assumption here that go's "int" type should always be copied over to a protobuf int64 type

Reply:
Is it true for a 32-bit build? Do you need to handle `int` and `uint` or could you stick to the fixed-width types?